### ISSUE TITLE
exception thrown from a function declared as noexcept

### ISSUE THREAD

**Issue by morin775** [msgpack/msgpack-c#1029]
The constructor of msgpack::zone has `noexcept` specifier

```c++
inline zone::zone(size_t chunk_size) noexcept:m_chunk_size(chunk_size), m_chunk_list(m_chunk_size)
{
}
```

but `m_chunk_list` in member initializer list can throw `std::bad_alloc`.

```c++
chunk_list(size_t chunk_size)
{
    chunk* c = static_cast<chunk*>(::malloc(sizeof(chunk) + chunk_size));
    if(!c) {
        throw std::bad_alloc();
    }

    m_head = c;
    m_free = chunk_size;
    m_ptr  = reinterpret_cast<char*>(c) + sizeof(chunk);
    c->m_next = MSGPACK_NULLPTR;
}
```

When an exception is thrown from a function declared as `noexcept`, `std::terminate` is called directly and the app crashes.
The user has no way to avoid the crash. Is it possible to catch the exception and report an error?

**Comment by redboltz** [comment]
Thank you for reporting the issue.
As you pointed out, chunk_list could throw the exception, so zone constructor shouldn't be noexcept.

I will create a PR to remove `noexcept` from `zone` constructor.

**Comment by redboltz** [comment]
Fixed by #1030

**Comment by morin775** [comment]
Thank you!
